manager/shard_strategy.go

```diff
@@ -0,0 +1,226 @@
+package manager
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"fmt"
+	"hash/fnv"
+
+	"github.com/flyteorg/flytepropeller/manager/config"
+	"github.com/flyteorg/flytepropeller/pkg/apis/flyteworkflow/v1alpha1"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+// Defines necessary functionality for a sharding strategy
+type ShardStrategy interface {
+	// Returns the total number of pods for the sharding strategy
+	GetPodCount() int
+	// Generates a unique hash code to identify shard strategy updates
+	HashCode() (uint32, error)
+	// Updates the PodSpec for the specified index to include label selectors
+	UpdatePodSpec(pod *v1.PodSpec, containerName string, podIndex int) error
+}
+
+// Creates and validates a new ShardStrategy defined by the configuration
+func NewShardStrategy(ctx context.Context, shardConfig config.ShardConfig) (ShardStrategy, error) {
+	switch shardConfig.Type {
+	case config.HashShardType:
+		if shardConfig.PodCount <= 0 {
+			return nil, fmt.Errorf("configured PodCount (%d) must be greater than zero", shardConfig.PodCount)
+		} else if shardConfig.PodCount > v1alpha1.ShardKeyspaceSize {
+			return nil, fmt.Errorf("configured PodCount (%d) is larger than available keyspace size (%d)", shardConfig.PodCount, v1alpha1.ShardKeyspaceSize)
+		}
+
+		return &HashShardStrategy{
+			EnableUncoveredReplica: shardConfig.EnableUncoveredReplica,
+			PodCount:               shardConfig.PodCount,
+		}, nil
+	case config.ProjectShardType, config.DomainShardType:
+		replicas := make([][]string, 0)
+		for _, replica := range shardConfig.Replicas {
+			if len(replica.Entities) == 0 {
+				return nil, fmt.Errorf("unable to create replica with 0 configured entity(ies)")
+			}
+
+			replicas = append(replicas, replica.Entities)
+		}
+
+		var envType environmentType
+		switch shardConfig.Type {
+		case config.ProjectShardType:
+			envType = project
+		case config.DomainShardType:
+			envType = domain
+		}
+
+		return &EnvironmentShardStrategy{
+			EnableUncoveredReplica: shardConfig.EnableUncoveredReplica,
+			EnvType:                envType,
+			Replicas:               replicas,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("shard strategy '%s' does not exist", shardConfig.Type)
+}
+
+// HashShardStrategy evenly assigns disjoint keyspace responsibilities over a collection of pods.
+// All FlyteWorkflows are assigned a shard-key using a hash of their executionID and are then
+// processed by the FlytePropeller instance responsible for that keyspace range.
+type HashShardStrategy struct {
```

Comment:
can we split each of the strategies into separate files?
One way would be to create a folder called `shard_strategy` and then under this folder
hash.go
env.go
etc?